src/compiler/checker.ts

```diff
@@ -7634,6 +7637,14 @@ namespace ts {
             return type;
         }
 
+        function containsThisProperty(thisProperty: Expression, expression: Expression) {
+            return isPropertyAccessExpression(thisProperty)
+                && thisProperty.expression.kind === SyntaxKind.ThisKeyword
+                && forEachChildRecursively(

```

Comment:
Would it be worth it to bail in a class expression / function expression? If you walk into a context that has a different `this`, I think the logic would be wrong. Also unclear whether it would be wrong in an arrow function expression in general, but probably would be right in an arrow function IIFE? 😬 At any rate, assigning functions that use `this` to a `this` property in JS code is a realistic test case, especially if we're talking about ES5-style constructor functions as opposed to classes.

Reply:
Briefly: no.

This logic could be wrong in two ways. It could return false when it should return true; it could miss a `this.x` reference nested on the rhs of a binary expression with a `this.x` lhs. Making this code more complex raises the risk of this happening; I can only think of it missing `this["x"]` right now. Missing a `this.x` reference produces (1) `any` (2) a bogus error, just like today.

Or it could return true when it should return false, which is the case you're pointing out. In this case, a single declaration incorrectly produces `any`, but other declarations will still provide the correct type, which will sometimes result in the right type being assigned to the property.

The penalty for being wrong isn't that bad, and the code is no longer so simple, so I don't think it's worth it, especially since no customers have complained that the *current* behaviour isn't correct.

Reply:
However, another test is a good idea to show the limits of the current code.

Reply:
@sandersn you can leverage `isMatchingReference` to at least be consistent with how control flow matches references, right?

Reply:
Yes, good idea.